lib/backend/model/keys.go

```diff
@@ -286,8 +286,22 @@ func ParseValue(key Key, rawData []byte) (interface{}, error) {
 		}
 	}
 	iface := value.Interface()
+
+retryWithEmptyJSON:
 	err = json.Unmarshal(rawData, iface)
 	if err != nil {
+		// This is a special case to address backwards compatibility from the time when we had no state information as block affinity value.
+		// example:
+		// Key: "/calico/ipam/v2/host/myhost.io/ipv4/block/172.29.82.0-26"
+		// Value: ""
+		// In 3.0.7 we added block affinity state as the value, so old "" value is no longer a valid JSON, so for that
+		// particular case we replace the "" with a "{}" so it can be parsed and we don't leak blocks after upgrade to Calico 3.0.7
+		// See: https://github.com/projectcalico/calico/issues/1956
+		if bytes.Equal(rawData, []byte(``)) && valueType == typeBlockAff {
+			rawData = []byte(`{}`)
+			goto retryWithEmptyJSON
```

Comment:
Rather than use a `goto` here, can't we just call `json.Unmarshal` here? I think that would be simpler and nicer.

Reply:
I thought of it but then thought it's more duplicate code, but I also don't like `goto` very much, so I'm fine either way. I'll change it